core/lib/presentation/views/pull_to_refresh/pull_to_refresh_widget.dart

```diff
@@ -0,0 +1,153 @@
+import 'dart:ui';
+
+import 'package:core/presentation/extensions/color_extension.dart';
+import 'package:core/presentation/utils/theme_utils.dart';
+import 'package:flutter/material.dart';
+import 'package:pull_to_refresh/pull_to_refresh.dart';
+
+class PullToRefreshWidget extends StatefulWidget {
+  final Widget child;
+  final Future<void> Function() onNormalRefresh;
+  final Future<void> Function() onDeepRefresh;
+  final double deepRefreshThreshold;
+  final String normalRefreshText;
+  final String deepRefreshText;
+  final String pullDownToRefreshText;
+  final String refreshingText;
+  final String pullingText;
+  final String releaseToText;
+  final String pullFurtherForText;
+  final String pullHarderForText;
+  final Icon? normalRefreshIcon;
+  final Icon? deepRefreshIcon;
+
+  const PullToRefreshWidget({
+    Key? key,
+    required this.child,
+    required this.onNormalRefresh,
+    required this.onDeepRefresh,
+    this.deepRefreshThreshold = 200.0,
+    this.normalRefreshText = 'Refresh',
+    this.deepRefreshText = 'Deep refresh',
+    this.pullDownToRefreshText = 'Pull down to refresh',
+    this.refreshingText = 'Refreshing',
+    this.pullingText = 'Pulling',
+    this.releaseToText = 'Release to',
+    this.pullFurtherForText = 'Pull further for',
+    this.pullHarderForText = 'Pull harder for',
+    this.normalRefreshIcon = const Icon(Icons.refresh, size: 20),
+    this.deepRefreshIcon = const Icon(Icons.autorenew, size: 20),
+  }) : super(key: key);
+
+  @override
+  State<PullToRefreshWidget> createState() => _PullToRefreshWidgetState();
+}
+
+class _PullToRefreshWidgetState extends State<PullToRefreshWidget> {
+  final RefreshController _refreshController = RefreshController();
+  double _scrollOffset = 0.0;
+  String? _lastAction;
+
+  @override
+  Widget build(BuildContext context) {
+    return NotificationListener<ScrollNotification>(
+      onNotification: (notification) {
+        if (notification is ScrollUpdateNotification) {
+          setState(() {
+            _scrollOffset = notification.metrics.pixels;
+          });
```

Comment:
Should we only call `setState` when pass certain threshold? To avoid rebuild the widget too many time unnecessarily.

Reply:
No, we must always update this value to check the current pull to refresh display state

Reply:
For example, `deepRefreshThreshold` is 200. If `_scrollOffset` is changing any where between 0-199, the deep search won't be activated. However, even 1 pixel change for `_scrollOffset` triggers `setState`. We can simply optimize this callback with:
```suggestion
          if (notification.metrics.pixels > widget.deepRefreshThreshold && _scrollOffset.abs() <= widget.deepRefreshThreshold) {
            setState(() {
              _scrollOffset = notification.metrics.pixels;
            });
          } else if (notification.metrics.pixels <= widget.deepRefreshThreshold && _scrollOffset.abs() > widget.deepRefreshThreshold) {
            setState(() {
              _scrollOffset = notification.metrics.pixels;
            });
          } 
```